thriftbp/client_middlewares_test.go

```diff
@@ -458,3 +459,103 @@ func TestSetClientName(t *testing.T) {
 		},
 	)
 }
+
+const (
+	localSvr        = "localsvr"
+	remoteSvr       = "remotesvr"
+	methodIsHealthy = "is_healthy"
+)
+
+func TestPrometheusClientMiddleware(t *testing.T) {
+	testCases := []struct {
+		name          string
+		wantErr       error
+		wantFail      bool
+		exceptionType string
+	}{
+		{
+			name:          "error",
+			wantErr:       errors.New("test"),
+			wantFail:      true,
+			exceptionType: "thrift.tApplicationException",
+		},
+		{
+			name:          "success",
+			wantErr:       nil,
+			wantFail:      false,
+			exceptionType: "",
+		},
+	}
+
+	for _, tt := range testCases {
+		t.Run(tt.name, func(t *testing.T) {
+			labelValues := []string{
+				localSvr,
+				methodIsHealthy,
+				strconv.FormatBool(!tt.wantFail),
+				tt.exceptionType,
+				"",
+				"",
+				remoteSvr,
+			}
+
+			requestLabelValues := []string{
+				localSvr,
+				methodIsHealthy,
+				remoteSvr,
+			}
+
+			defer thriftbp.PrometheusClientMetricsTest(t, labelValues, requestLabelValues).CheckMetrics()
+
+			ctx := context.Background()
+			handler := mockBaseplateService{fail: tt.wantFail, err: tt.wantErr}
+			client := setupFake(ctx, t, handler)
+			bpClient := baseplatethrift.NewBaseplateServiceV2Client(client.TClient())
+			result, err := bpClient.IsHealthy(
+				ctx,
+				&baseplatethrift.IsHealthyRequest{
+					Probe: baseplatethrift.IsHealthyProbePtr(baseplatethrift.IsHealthyProbe_READINESS),
+				},
+			)
+			if tt.wantErr != nil && err == nil {
+				t.Error("expected an error, got nil")
+			} else if tt.wantErr == nil && err != nil {
+				t.Errorf("expected no error, got %v", err)
+			}
+
+			if result == tt.wantFail {
+				t.Errorf("result mismatch, expected %v, got %v", tt.wantFail, result)
+			}
+		})
+	}
+}
+
+type mockBaseplateService struct {
+	fail bool
+	err  error
+}
+
+func (srv mockBaseplateService) IsHealthy(ctx context.Context, req *baseplatethrift.IsHealthyRequest) (r bool, err error) {
+	return !srv.fail, srv.err
+}
```

Comment:
@kylelemons or @fishy (or anyone else): For readability, I would like to be explicit about when I'm implementing an interface.  In the past I've done this:
```
var _ baseplatethrift.BaseplateServiceV2 = (*mockBaseplateService)(nil)
```
But I'm reading in Effective Go "Don't do this for every type that satisfies an interface" [here](https://golang.org/doc/effective_go#blank_implements).

Also in Effective Go [docs](https://golang.org/doc/effective_go#generality), it describes "if a type exists only to implement an interface and will never have exported methods beyond that interface, there is no need to export the type itself. Export just the interface...", but this case doesn't apply to all my times when I want to be explicit  about interfaces I've implemented. 

What about adding this?
```
var _ baseplatethrift.BaseplateServiceV2 = mockBaseplateService{}
```

Or should I simply not try to be explicit about when an interface is implemented?

Reply:
Reading through that part of Effective Go, the next bit it says is:

>  By convention, such declarations are only used when there are no static conversions already present in the code, which is a rare event.

The main goal of blocks like that IMO is not to say what interface something implements, but to verify that it implements it. In this case, your code wouldn't compile if `mockBaseplateService` didn't implement `baseplatethrift.BaseplateServiceV2` so you don't really need that check (because you pass it to `setupFake` which expects handler to implement that interface).

Reply:
@pacejackson thanks for the response. I get all that and I understand that the compiler will enforce it in this case of  `mockBaseplateService`, but I'm more asking in general. 

My question is more with regards to readability for the programmer ... I want to communicate to the reader that the object they are looking at implements a specific interface.  That make sense? I like how java has an "implements" keyword because its explicit. I know golang interfaces are not explicit, but just curious if there is a way to make them so because I like it more for readability.

Reply:
Personally, I don't think it actually adds to the readability. I feel like it's usually pretty clear when you're implementing an interface.